Propagating max connections when using DataSourceJdbcDataSource

renatocaval opened the pull request:
Fixes #1814

TimMoore commented:
A test failed:

```
[error] Test slick.test.jdbc.DataSourceTest.testConnectionPoolDisabled failed: java.lang.AssertionError: maxConnections should be None when not using a pool expected:<None> but was:<Some(100)>, took 0.004 sec
[error]     at slick.test.jdbc.DataSourceTest.testConnectionPoolDisabled(DataSourceTest.scala:151)
[error]     ...
```

This implies the current behavior is deliberate... @szeiger or another Slick maintainer can you comment?

TimMoore commented on slick/src/main/scala/slick/jdbc/JdbcDataSource.scala:
```diff
@@ -99,7 +99,9 @@ object DataSourceJdbcDataSource extends JdbcDataSourceFactory {
     }
-    new DataSourceJdbcDataSource(ds, c.getBooleanOr("keepAliveConnection"), None, new ConnectionPreparer(c))
+    val numThreads = c.getIntOr("numThreads", 20)
+    val maxConnections = Some(c.getIntOr("maxConnections", numThreads * 5))
```
This is incorrect in the case where `dataSourceClass` is `None`: `maxConnections` should be too. Could move into the match statement above `val (ds, maxConnections) = ...`

renatocaval replied:
Yes, but in any case we should also not try to come up with a value for `maxConnections` like I tried. 

When we do have a `dataSourceClass` defined, the `maxConnections` should be:
`val maxConnections = c.getIntOpt("maxConnections")`

renatocaval commented:
Tests failed for no apparent reason. Flakiness?

d6y commented:
@renatocaval looks like it was a flakey db2 test - I kicked it off again, and passed this time.

renatocaval commented:
Ok, thanks. 

I'm planning to work further on this PR. I will add a WIP on the title.

renatocaval commented:
@d6y I may add some extra test. Something for tomorrow. ;-)

hvesalai commented on slick/src/main/scala/slick/jdbc/JdbcBackend.scala:
```diff
@@ -287,7 +294,7 @@ trait JdbcBackend extends RelationalBackend {
       val poolName = usedConfig.getStringOr("poolName", path)
       val numThreads = usedConfig.getIntOr("numThreads", 20)
-      val maxConnections = source.maxConnections.fold(numThreads*5)(identity)
+      val maxConnections = source.maxConnections.getOrElse(numThreads*5)
```
Changed in #1855 to be `numThreads`, without `* 5` because having more connections than threads means a connection without a thread can be holding a lock on the db, preventing all of the threads from continuing.

renatocaval replied:
Ok, correct. 
I'm however updating this line. A `fold` on an `Option` using `identity` is a overkill and also obscure. `getOrElse` should be preferred.

renatocaval commented:
@hvesalai @d6y I removed the WIP from the title. Rebase with master and squashed the commits. 
This is ready for final review and merge if you guys agree. 

There is another issue that I wanted to fix, but I realise that it's much more complicated and therefore I will open a new issue and we can discuss possible solutions there.

hvesalai commented:
LGTM. Will merge after CI success.

hvesalai commented:
Travis is having issues at the moment, so I will restart the build later.

renatocaval commented:
FYI, here is the new issue I created, #1859. I'll send a new PR (incomplete) so we can further discuss possible fixes.

cvogt commented:
LGTM
